### PR TITLE
React-based prizes index and detail views

### PR DESCRIPTION
**Pull request by faultyserver**
- [ ] I've added tests or modified existing tests for the change.
- [X] I've humanly end-to-end tested the change by running an instance of the tracker.

(more tests pending merge of #141 for infra, ideally non-blocking on this PR)

### Issue from Pivotal Tracker

https://www.pivotaltracker.com/story/show/169234618

### Description of the Change

This PR revamps the existing prize pages (e.g., https://gamesdonequick.com/tracker/prizes/GDQX2019) into React, along with a general design overhaul to make them feel more modern, well-kept, and friendly.

Here are some example uses that show the new look and flow:

|mobile|desktop|
|-------|--------|
| image|image|
|image|image|
|image|image|

(note: these examples came before fixing an issue with the Closing Soon calculation.

Included in this PR:

- A new Prizes index containing a grid of Prizes with a cover image, name, availability, and basic details to make searching through prizes more enjoyable and effective.
- A new Prize Details page showcasing an image of the prize and all of the information about it. (includes controlled Markdown support for prize descriptions!)
- Initial Best Practices examples for data loading, fetching, presentation, and routing.
- Standardized API calls to fetch and load data into stores.
- Support for pre-filling the Donation form with an amount via the url's hash.
- Lots of infra changes! (see below)

Not included in this PR are:

- Highlighting for grand prizes on the index page (hopefully coming soon).
- Pagination of the index to avoid loading too much data at once (browsers are okay at handling this, though)
- Filtering of prizes on the index to find specific prizes (ctrl+f does this just fine, but better filtering with fuzzysearch, run names, etc. would be cool)
- Alternate images for prize details (we have 3 files sometimes available. Only one is canonical, but adding more could be nice long term/for future events)

This migration involved a lot of additional changes for the frontend, including:

- renaming the `donate` bundle to `tracker`, since it now includes more than just the donate page.
- Adding `TimeUtils` (currently powered by `luxon`) for easily handling time information.
- Adding `react-router` and it's necessities to provide routing between all of the now-React pages.
- Refactoring `Anchor` to work with both local URLs via react-router and external URLs natively.
- Implementing a "translation layer" in actions to load API data into a frontend-friendly format, and to handle parsing composite types like dates and currencies.
- A "fake" implementation of search filters based on some very-surface-level understanding of how it works and what it accepts.

### Possible Drawbacks

Some behavior like using event shortnames for the eventId in the url are not currently supported. Given the router structure, this _should_ be fairly easy to make work, but I think there should be some more well-contained infrastructure for the sake of efficient loading and to avoid decentralizing the responsibility of knowing how the event was specified.

The bundle for the app is also starting to approach a large enough size for code splitting to be effective, but I think it's still more than reasonable for now:

```
# production build of this branch
tracker (725 KiB)
      tracker-tracker-c0c743a1c206daebe5b3.css
      tracker-tracker-c0c743a1c206daebe5b3.js
```

I wouldn't recommend adding codesplitting until we have more than like 8-10 routes. `uikit` a large part of the size is currently shared on all pages, so the actual benefit would be minimal.

### Verification Process

Constant navigation through pages, dealing with both local data and mock datas from API calls on the production GDQ server.

### CODE REVIEW

File: bundles/public/util/currency.ts [hunk 1]
```diff
@@ -7,3 +7,14 @@ export function parseCurrency(amount?: string) {
   const parsed = parseFloat(amount);
   return Number.isNaN(parsed) ? undefined : parsed;
 }
+
+// Like `parseCurrency`, but enforces that the given amount exists, and that it
+// is parsed into an actual number (and not NaN).
+export function parseCurrencyForced(amount: string) {
+  const parsed = parseFloat(amount);
+  if (Number.isNaN(parsed)) {
+    throw new TypeError(`String value "${amount}" could not be parsed as a currency`);
```
Comment (justinkim) [inline bundles/public/util/currency.ts]: so what happens here, does it break the page if we get here?

File: bundles/tracker/events/EventConstants.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+// TODO: Have this be set per-event
+export const SWEEPSTAKES_URL = 'https://gamesdonequick.com/sweepstakes';
```
Comment (justinkim) [inline bundles/tracker/events/EventConstants.ts]: any way to pull this out into a config file?

File: bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+// times when they are within a certain range. Too far out and the schedule is
+// likely to change before the time window arrives. Too close and we can't be
+// sure that the displayed time is accurate enough.
+//
+// These numbers configure the furthest and closest times that we will show
+// relative times in the description of the card. They are mostly arbitrary.
+const ALLOWED_ESTIMATED_TIMES = [
+  { time: 15 * 60 * 1000, display: 'in about 15 minutes' },
+  { time: 30 * 60 * 1000, display: 'in about 30 minutes' },
+  { time: 45 * 60 * 1000, display: 'in about 30 minutes' },
```
Comment (justinkim) [inline bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx]: 45 minutes

File: bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+// closing at 11am on day 2 will show `Closes <day 2>`.
+const EXACT_TIME_RELATIVE_LIMIT = 11 * 60 * 60 * 1000; // 11 hours
+
+// Returns a string representation of the relative duration given as `millis`.
+// For example, 8 minutes (8 * 60 * 1000 millis) would return the string
+// "about 15 minutes" with the default configuration.
+//
+// If the given duration is longer than the maximum allowed relative time,
+// or if it is less than or equal to 0, `undefined` is returned instead.
+function getDisplayableEstimatedTime(now: DateTime, millis: number): string | undefined {
```
Comment (justinkim) [inline bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx]: Can we get tests for this and `getDisplayableExactTime`

File: bundles/public/util/currency.ts (hunk 1 above)
Comment (faultyserver) [reply bundles/public/util/currency.ts]: correct. I should add a comment that this is not a practical utility and is only meant for parsing API data that is known to be in a given format, as is also done with datetimes.

The usage pattern already catches that when any error with _fetching_ the response happens, it is handled safely, so the transformation code can (almost 100%) safely assume that the API response it's parsing is well-formed.

If it's not, there can't really be any guarantees about page performance, so it _should_ crash, and our error boundary catches it. What we should then do is build a production-quality error boundary that mentions the crash instead of rendering blank.

File: bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx (hunk 3 above)
Comment (faultyserver) [reply bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx]: rip ty

File: bundles/tracker/router/components/NotFound.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+
+import Container from '../../../uikit/Container';
+import Header from '../../../uikit/Header';
+import Text from '../../../uikit/Text';
+
+const NotFound = () => {
+  return (
+    <Container size={Container.Sizes.NORMAL}>
+      <Header>That page does not exist</Header>
+      <Text>Sorry about that.</Text>
```
Comment (justinkim) [inline bundles/tracker/router/components/NotFound.tsx]: :(

File: bundles/tracker/events/EventConstants.ts (hunk 2 above)
Comment (faultyserver) [reply bundles/tracker/events/EventConstants.ts]: This shouldn't exist in general because it should be set per-event. Beyond that, the intention of these Constants files is to be configurable for the code that consumes it, but it is scoped to individual modules for the sake of avoiding a 3k line long mess of constants.

tl;dr: I would consider this the config file for a front end project. If a separate consumer needs to change this before it becomes an event field, this is the place to look. Global config gets messy.

File: bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx (hunk 4 above)
Comment (faultyserver) [reply bundles/tracker/prizes/components/PrizeRelativeAvailability.tsx]: I'm waiting for #141 to add tests because that has some infra that I don't want to re-write in the meantime.

This does need tests eventually though.

File: bundles/tracker/router/components/NotFound.tsx (hunk 5 above)
Comment (faultyserver) [reply bundles/tracker/router/components/NotFound.tsx]: I'd like to have a "Take me back to Home" button or something, but there isn't a canonical "home" route handled by React atm, so it doesn't make sense.

File: bundles/tracker/events/EventConstants.ts (hunk 2 above)
Comment (justinkim) [reply bundles/tracker/events/EventConstants.ts]: Ok, if this is going to be the config file, let's make mention of it in the README so users of the tracker can know about it.

Comment (faultyserver) [reply bundles/tracker/events/EventConstants.ts]: I'm 100% for making it clear how to adapt this project for different consumers, but I'm hesitant that just pointing to this file is the right thing to do. Namely because there are (and will be more) files for config such as https://github.com/faultyserver/donation-tracker/blob/feature/prizes-react/bundles/tracker/donation/DonationConstants.ts that relate to their respective modules.

What I _really_ want to do is have a full front-end guide that explains a bunch of things like theming, various features of `uikit`, the module system in general, routing, external links (like this), and general best-practices for modifying and building new features. My goal is to do that during or right after AGDQ.

Comment (uraniumanchor) [reply bundles/tracker/events/EventConstants.ts]: I have some thoughts about how to have a generic way to load this sort of global config into a React context at the top level but it can wait until after the event.

File: bundles/tracker/prizes/components/Prize.tsx [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        <Text>
+          No donation necessary for a chance to win. See <Anchor href={SWEEPSTAKES_URL}>sweepstakes rules</Anchor> for
+          details and instructions.
+        </Text>
+      </div>
+    </Container>
+  );
+};
+
+export default Prize;
```
Comment (uraniumanchor) [inline bundles/tracker/prizes/components/Prize.tsx]: Not sure where it is in this file but when I try to view a prize card it spits this proptype warning in the console:

```
backend.js:6 Warning: Failed prop type: The prop `to` is marked as required in `Link`, but its value is `undefined`.
    in Link (created by Anchor)
    in Anchor (created by Prize)
    in span (created by Prize)
    in p (created by Text)
    in Text (created by Prize)
    in div (created by Prize)
    in div (created by Prize)
    in div (created by Prize)
    in div (created by Container)
    in Container (created by Prize)
    in Prize (created by Context.Consumer)
    in Route (created by EventRouter)
    in Switch (created by EventRouter)
    in EventRouter (created by Context.Consumer)
    in Route (created by App)
    in Switch (created by App)
    in Router (created by App)
    in App
    in ErrorBoundary
    in div (created by ThemeProvider)
    in ThemeProvider
    in Provider
```

Comment (uraniumanchor) [reply bundles/tracker/prizes/components/Prize.tsx]: Also, the image does not seem to be showing up, but shows up on the carousel.

image

``

image

Comment (uraniumanchor) [reply bundles/tracker/prizes/components/Prize.tsx]: Here's the response I have locally:

```
[{"model": "tracker.prize", "pk": 1, "fields": {"name": "Test Prize", "category": null, "image": "", "altimage": "", "imagefile": "/media/prizes/20191126_093545_vRN06dk.jpg", "description": "", "shortdescription": "", "extrainfo": "", "estimatedvalue": null, "minimumbid": "10.00", "maximumbid": "10.00", "sumdonations": false, "randomdraw": true, "event": 1, "startrun": 1, "endrun": 1, "starttime": null, "endtime": null, "maxwinners": 1, "maxmultiwin": 1, "provider": "", "handler": 1, "acceptemailsent": false, "creator": null, "creatoremail": null, "creatorwebsite": null, "state": "PENDING", "requiresshipping": true, "reviewnotes": "", "custom_country_filter": false, "key_code": false, "allowed_prize_countries": [], "disallowed_prize_regions": [], "start_draw_time": "2019-11-02T16:05:00Z", "end_draw_time": "2019-11-02T17:20:00Z", "canonical_url": "http://localhost:8000/tracker/prize/1", "public": "Test Prize", "numwinners": "0", "startrun__name": "Foo3", "startrun__display_name": "Foo", "startrun__twitch_name": "", "startrun__deprecated_runners": "Foo", "startrun__console": "", "startrun__commentators": "", "startrun__description": "", "startrun__starttime": "2019-11-02T16:10:00Z", "startrun__endtime": "2019-11-02T16:20:00Z", "startrun__order": 2, "startrun__run_time": "0:05:00", "startrun__setup_time": "0:05:00", "startrun__coop": false, "startrun__category": null, "startrun__release_year": null, "startrun__tech_notes": "", "startrun__public": "Foo3 (Foo)", "endrun__name": "Foo3", "endrun__display_name": "Foo", "endrun__twitch_name": "", "endrun__deprecated_runners": "Foo", "endrun__console": "", "endrun__commentators": "", "endrun__description": "", "endrun__starttime": "2019-11-02T16:10:00Z", "endrun__endtime": "2019-11-02T16:20:00Z", "endrun__order": 2, "endrun__run_time": "0:05:00", "endrun__setup_time": "0:05:00", "endrun__coop": false, "endrun__category": null, "endrun__release_year": null, "endrun__tech_notes": "", "endrun__public": "Foo3 (Foo)"}}]
```

File: ui/urls.py [hunk 7]
```diff
@@ -6,4 +6,7 @@
     url(r'^$', views.index, name='index'),
     url(r'^admin/', views.admin, name='admin'),
     url(r'^donate/(?P<event>\w+)$', views.donate, name='donate'),
+    url(r'^prizes/$', views.index, name='prizes'),
+    url(r'^prizes/(?P<prize>\w+)$', views.index, name='prize'),
+    url(r'^events/', views.index, name='events'),
```
Comment (uraniumanchor) [inline ui/urls.py]: These don't seem to match what the React router wants, so I was a bit confused trying to get this working initially.

Comment (faultyserver) [reply ui/urls.py]: oh, I think this was from before I settled on routes on the front. Only `events/` needs to be here now I guess.

File: bundles/tracker/prizes/components/Prize.tsx (hunk 6 above)
Comment (faultyserver) [reply bundles/tracker/prizes/components/Prize.tsx]: This isn't related but:

```
"imagefile": "/media/prizes/20191126_093545_vRN06dk.jpg"
```

shouldn't this be a full url and not just a path?

This issue I think is because the prize page doesn't do the same resolution logic that the cards do. Can fix soon

Comment (faultyserver) [reply bundles/tracker/prizes/components/Prize.tsx]: And Link is just me setting `href` instead of `to`. ez fix.

Comment (uraniumanchor) [reply bundles/tracker/prizes/components/Prize.tsx]: > shouldn't this be a full url and not just a path?

Right now it's just whatever the storage backend says the url is. When I did this, I'm using the default "store files directly on the server" backend, with `MEDIA_URL` set to just `/media/`, which is a valid way to set it up, I'm pretty sure.

Comment (uraniumanchor) [reply bundles/tracker/prizes/components/Prize.tsx]: Reference:

https://docs.djangoproject.com/en/1.11/ref/settings/#media-url
https://docs.djangoproject.com/en/1.11/howto/static-files/#serving-uploaded-files-in-development

Comment (faultyserver) [reply bundles/tracker/prizes/components/Prize.tsx]: I see. nbd for now, just confused myself from the discussion on #147. Doesn't affect how the string gets used so it's all fine.
